front/src/common/RollingStockSelector/RollingStockModal.tsx

```diff
@@ -49,65 +92,39 @@ function RollingStockModal(props) {
     }
   );
 
-  const searchMateriel = (e) => {
+  const searchMateriel = (e: React.ChangeEvent<HTMLInputElement>) => {
     setFilters({ ...filters, text: e.target.value.toLowerCase() });
-    setIsFiltering(true);
+    setIsLoading(true);
   };
 
   const updateSearch = () => {
     setOpenedRollingStockCardId(undefined);
-
-    const filterText = filters.text;
-    const filterIsNotEmpty = filterText || filters.elec || filters.thermal;
-
-    const filteredRollingStocks = filterIsNotEmpty
-      ? rollingStocks.filter((rollingStock) => {
-          // checkbox filters
-          if (filters.elec || filters.thermal) {
-            const effortCurveModes = Object.values(rollingStock.effort_curves.modes).map(
-              // eslint-disable-next-line camelcase
-              ({ is_electric }) => is_electric
-            );
-            const isElectric = effortCurveModes.includes(true);
-            const isThermal = effortCurveModes.includes(false);
-            if ((filters.elec && !isElectric) || (filters.thermal && !isThermal)) {
-              return false;
-            }
-          }
-
-          // text filter
-          if (filterText) {
-            const { metadata } = rollingStock;
-            const containsFilterText = some(
-              [
-                rollingStock.name,
-                metadata.detail,
-                metadata.reference,
-                metadata.series,
-                metadata.type,
-                metadata.grouping,
-              ],
-              (string) => string && string.toLowerCase().includes(filterText)
-            );
-            if (!containsFilterText) {
-              return false;
-            }
-          }
-
-          // the rolling stock has passed both the checkbox filter and the text filter
-          return true;
-        })
-      : rollingStocks;
-
-    setTimeout(() => {
-      setFilteredRollingStockList(filteredRollingStocks);
-      setIsFiltering(false);
-    }, 0);
+    const newFilteredRollingStock = rollingStocks?.filter(
+      ({ name, metadata, effort_curves: effortCurves }) => {
+        const passSearchedStringFilter = rollingStockPassesSearchedStringFilter(
+          name,
+          metadata,
+          filters
+        );
+        const passEnergeticModesFilter = rollingStockPassesEnergeticModeFilters(
+          filters.elec,
+          filters.thermal,
+          effortCurves.modes
+        );
+        return passSearchedStringFilter && passEnergeticModesFilter;
+      }
+    );
+    if (newFilteredRollingStock) {
```

Comment:
```suggestion
```
No need for this test as newFilteredRollingStock is necessarily a LightRollingStock[]